### ISSUE TITLE
Collapse cell brings focus to the end of collapse section

### ISSUE THREAD

**Issue by ramanova** [jupyterlab/jupyterlab#8252]
## Description

After collapsing a long cell output, the focus stays in coordinates related to the previous collapsed section instead of the cell that this output belongs to.

## Reproduce

1. Go to a cell with long output
2. Scroll through the output, go to a middle e.g.
3. Press collapse ribbon on the left
4. See that the focus is computed by this absolute coordinate, now you are at some random code location instead of the original cell, that produced the output

**Comment by jasongrout** [comment]
This fix for this is likely similar to https://github.com/jupyterlab/jupyterlab/pull/9400 or https://github.com/jupyterlab/jupyterlab/pull/9273

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/minor-ux-collapsing-very-tall-output/7495/5

**Comment by gereleth** [comment]
I tried to take a stab at this.

I looked at the [`packages/notebook/src/actions.tsx > hideOutput`](https://github.com/jupyterlab/jupyterlab/blob/105c72631e5d7c248a3cbee8024e880410b51685/packages/notebook/src/actions.tsx#L1202) function. Turns out it already has correct scrolling behaviour. So if you scroll down a long output of a selected cell, then go to the menu and click "View > Collapse Selected Output" then you are scrolled correctly to the cell start.

However, this function doesn't get called when the collapse ribbon is clicked. What happens there is [`packages/cells/src/collapser.tsx > handleClick`](https://github.com/jupyterlab/jupyterlab/blob/105c72631e5d7c248a3cbee8024e880410b51685/packages/cells/src/collapser.tsx#L134). The click handler basically just toggles `cell.outputHidden` property.

I'm pretty new to jupyter lab and I don't know how to call the correct `hideOutput` function from the click handler. It needs a `notebook` and there's only a `cell` in the click handler.

**Comment by krassowski** [comment]
Thank you for looking into it @gereleth. I had a quick look and realised that not all cells belong to notebooks - some cells are displayed in the console instead and some are even used as standalone cells in debugger dialog window. Also, the `@jupyterlab/cells` does not depend on `@jupyterlab/notebook`, but the `@jupyterlab/notebook` depends on `@jupyterlab/cells`.

An easy way out would be to try to see what is in `cell.parent` (and possibly `.parent` again? this needs some experimenting, I am not sure) and if you can find a notebook then you have it. However, this might introduce a dependency on notebook in `@jupyterlab/cells`, which we may want to avoid.

Maybe instead we should modify how hiding cell output works: a cell would emit a signal when its output gets hidden and the notebook would listen on this signal to perform scroll if needed. Does it make sense?

**Comment by gereleth** [comment]
Importing from `@jupyterlab/notebook` in the `cells` package gives multiple ts errors like `error TS5055: Cannot write file '.../jupyterlab/packages/cells/lib/collapser.d.ts' because it would overwrite input file.` I guess it doesn't like the circular dependency between `notebook` and `cells`. So that way is probably out.

Emitting/handling signals makes sense, but I'm not sure how to implement that.

There's a third option I think - since in the click handler we have a `cell` we could scroll it into view with `ElementExt.scrollIntoViewIfNeeded`. If I add the following code in the collapser click handler then I get the expected behaviour when collapsing a long output.

```ts
protected handleClick(e: React.MouseEvent<HTMLDivElement>): void {
  const cell = this.parent?.parent as CodeCell | undefined | null;
  if (cell) {
    cell.outputHidden = !cell.outputHidden;
    // +++ scroll cell into view after output collapse
    if (cell.outputHidden) {
      const area = cell.parent?.node
      if (area) {
        ElementExt.scrollIntoViewIfNeeded(area, cell.node)
      }
    }
    // +++
  }
  /* We need this until we watch the cell state */
  this.update();
}
```

I have doubts about this solution because it's probably not the cell's responsibility to control scrolling in a notebook. 
On the other hand, maybe the cell **is** responsible for staying visible while the user is directly interacting with it.

I also haven't checked yet how this affects non-notebook cells (like those in console and debugger).

@krassowski do you think it's better to go for signals or to scroll directly in the collapser code?

**Comment by krassowski** [comment]
It's a good idea, I like it. On one hand, we already use this approach of `scrollIntoViewIfNeeded` when scrolling in debugger, on the other I am not sure how it would interact with the virtual notebook (in short the idea is that we will soon only render a fraction of cells that are currently visible to improve performance of rendering long notebooks).

The idea of using `scrollIntoViewIfNeeded` on a cell is also better because it does not introduce any new API (the signal would need to be public) and we won't have to worry about making it future-proof, and it should be context agnostic (should work in notebook, console, and debugger).

**Comment by krassowski** [comment]
Fixed in #10491 thanks to @gereleth, it will be available in 3.1 :tada:. Here is how it looks:

![scroll-ok](https://user-images.githubusercontent.com/5832902/124333603-9d8b7980-db8c-11eb-95f5-74be7f736de6.gif)
